datadog_checks_base/datadog_checks/base/utils/concurrency/limiter.py

```diff
@@ -0,0 +1,72 @@
+# (C) Datadog, Inc. 2022-present
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+import threading
+
+
+class ConditionLimiter(object):
+    """
+    This class is used to limit the number of concurrent satisfied conditions. Every subclass must implement
+    a `condition` method. For example, if you wanted to only run one check instance at a time based on the
+    command line arguments used to start an active process, you could do:
+
+    ```python
+    import re
+    import shlex
+
+    import psutil
+
+    from datadog_checks.base import AgentCheck
+    from datadog_checks.base.utils.concurrency.limiter import ConditionLimiter
+
+    class Limiter(ConditionLimiter):
+        def condition(self, pattern, logger):
+            logger.info('Searching for a process that matches: %s', pattern)
+            for process in psutil.process_iter(['cmdline']):
+                command = shlex.join(process.info['cmdline'])
+                if re.search(pattern, command):
+                    logger.info('Process found: %s', command)
+                    return True
+            else:
+                logger.info('Process not found, skipping check run')
+                return False
+
+    class Check(AgentCheck):
+        limiter = Limiter()
+
+        def check(self, _):
+            if not self.limiter.check_condition(self.check_id, self.config.pattern, self.log):
+                return
```

Comment:
> Does this mean that some check instances might never run?

Yes, which is desirable/required in some scenarios.

Reply:
In the case of IBM MQ (the literal example), only one configured instance will be successful because only one queue manager will be running in HA setups. The others would all fail as expected.

Reply:
So the idea that each check instance would first verify it can talk to the queue manager before running, and we cache this information until an instance that won the limit explicitly releases it? 

If yes, how would this interact with auto-discovery? What if the check instance that holds the slot is dynamically removed? (also see https://github.com/DataDog/datadog-agent/pull/13777)

Reply:
Added `cancel` to example